### PR TITLE
WIP: Tanaya/umbrella

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
WIP: UMBRELLA SAMPLING 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Samplers/Umbrella.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+      @ In, None
+      @ Out, None
+    """
+    ForwardSampler.__init__(self)
+    self.printTag = 'SAMPLER UMBRELLA'
+    self.samplingType = None
+    self.limit = None
+    self.modeLocation = 0
+    self.rho = 0
+    self.dimension = 0
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:91]: Variables introduced in the constructor should have a comment after them describing the variable; for example, "rho" is not clear unless the dev is familiar with the literature for the method

File: framework/Samplers/Umbrella.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+  def localInputAndChecks(self,xmlNode, paramInput):
+    """
+      Class specific xml inputs will be read here and checked for validity.
+      @ In, xmlNode, xml.etree.ElementTree.Element, The xml element node that will be checked against the available options specific to this Sampler.
+      @ In, paramInput, InputData.ParameterInput, the parsed parameters
+      @ Out, None
+    """
+    ForwardSampler.readSamplerInit(self,xmlNode)
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:100]: we're migrating towards using `super` instead, as:
```python
super().readSamplerInit(xmlNode)
```

File: framework/Samplers/Umbrella.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        self.samplingType = None
+    else:
+      self.raiseAnError(IOError,self,'Umbrella sampler '+self.name+' needs the samplerInit block')
+
+  def getStd(self, k, sigma):
+    """
+         Method to get alpha and beta values of gamma distribution based on mode location, standard deviation and tail probability
+         @ In, k, mode location of the gamma distribution specified by the user
+         @ In, tail, tail probability of the distribution
+         @ Out, sigma, standard distribution of the gamma distribution
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:124]: The docstrings don't match the method arguments here; this is true for many docstrings added in this PR; I'll just mention it here.

Comment (PaulTalbot-INL) [reply framework/Samplers/Umbrella.py:124]: Also, the format for the in/out docstrings are missing the "type". The format is
```python
"""
  @ In, varName, varType, description
"""
```
such as
```python
"""
  @ In, age, int, age of the unit used in this calculation
"""
```

File: framework/Samplers/Umbrella.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    return (scipy.stats.gamma.cdf(k,
+                                  a=alpha,
+                                  scale=1 / beta) - (1- self.tailProb))
+
+  def tailProbabilityGamma(self, k, sigma):
+    """
+         Method to collect parameters of gamma distribution in a dictionary
+         @ In, k, mode location of the gamma distribution specified by the user
+         @ In, sigma, standard distribution of the gamma distribution
+         @ Out, dictionary of gamma parameters
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:137]: The variable types are missing.

Also, Out arguments are also required to have names; if there's not a variable name that tracks the output, the name is the same as the function name. In this case,
```python
"""
@ Out, tailProbabilityGamma, dict, dictionary of gamma parameters
"""
```

File: framework/Samplers/Umbrella.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    """
+    D = k ** 2 + 4 * sigma ** 2
+    beta = (k + D ** 0.5) / (2 * sigma ** 2)
+    alpha = beta * k + 1
+    value = 1 - gamma.cdf(k, scale=1 / beta, a=alpha)
+    return {'alpha': alpha, 'beta': beta, 'tailprob': value,
+            'k': k, 'sigma': sigma}
+
+  def getDiscreteSamples(self, n, the_w):
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:146]: underscores are not allowed in naming variables in RAVEN

File: framework/Samplers/Umbrella.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+         @ In, dimensions, dimensions specified by the user
+         @ Out, dataframe, vertices in K dimensional unit square
+    """
+    nrows = 2 ** self.dimension
+    vertices = np.zeros(shape=(nrows, self.dimension))
+    for i in range(1, nrows + 1):
+      for j in range(1, self.dimension + 1):
+        power = math.floor((i - 1) / 2 ** (j - 1))
+        vertices[i - 1, j - 1] = (-1) ** (power + 2)
+    return vertices
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:174]: This may be more applicable as a `mathUtils` function than a method here

File: framework/Samplers/Umbrella.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    """
+    sample = gamma.rvs(scale=1 / self.tailProb ['beta'],
+                       a=self.tailProb ['alpha'],
+                       size=sample_size)
+    return sample
+
+  def gammaTailSample(self, sampleSize, vertWeights):
+    """
+         Method to generate multivariate gamma tail samples
+         @ In, sample_size
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:233]: docstring doesn't match arguments

File: framework/Samplers/Umbrella.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    newSamplesDF = pd.DataFrame(newSamples)
+    newSamplesDF['vert'] = vertexSample
+    newSamplesDF['tail_flag'] = 1
+
+    return newSamplesDF
+
+  def umbrellaSample(self, sampleSize, vertWeights):
+    """
+         Method to generate samples from multivariate umbrella distribution
+         @ In, sample_size
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:256]: variable name mismatch

File: framework/Samplers/Umbrella.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+      self.covariance[i, i] = 1
+
+    sigma = scipy.optimize.root(partial(self.getStd, self.modeLocation), 1, tol=0.0000001)['x'][0]
+    self.tailProb = self.tailProbabilityGamma(self.modeLocation, sigma)
+    sampleSize = 30
+    umbrellaSample = self.umbrellaSample(sampleSize,
+                                         verticesWeights)
+
+    print('umbrellaSample')
+    print(umbrellaSample)
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Umbrella.py:306]: this prints should be removed before merging

File: tests/framework/ROM/TimeSeries/ARMA/basic.xml [hunk 10]
```diff
@@ -74,8 +74,11 @@
       <Features>scaling</Features>
       <pivotParameter>Time</pivotParameter>
       <outTruncation domain="positive">Speed</outTruncation>
-      <Fourier>604800, 302400, 86400, 43200</Fourier>
-      <P>2</P>
+		<!--<Fourier>604800, 86400, 43200, 3600, 1800</Fourier>-->
+      <Fourier>2678400, 1339200,  892800,  669600, 535680,  446400</Fourier>
```
Comment (PaulTalbot-INL) [inline tests/framework/ROM/TimeSeries/ARMA/basic.xml:78]: why are we changing a base ARMA test here?

File: tests/framework/test_Importance_Sampler.xml [hunk 11]
```diff
@@ -0,0 +1,98 @@
+<?xml version="1.0" ?>
```
Comment (PaulTalbot-INL) [inline tests/framework/test_Importance_Sampler.xml:1]: I think this test would be better placed in `raven/tests/framework/Samplers/`

File: tests/framework/test_Umbrella_Sampler.xml [hunk 12]
```diff
@@ -0,0 +1,97 @@
+<?xml version="1.0" ?>
```
Comment (PaulTalbot-INL) [inline tests/framework/test_Umbrella_Sampler.xml:1]: this would also be better in `raven/tests/framework/Samplers`

**Review by PaulTalbot-INL — COMMENTED** [review]
A few initial comments as we continue iterating on this. It's getting close; Tanaya did a lot of cleaning up, so there's just a little adjusting and testing to do.

File: framework/Samplers/Importance.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from .ForwardSampler import ForwardSampler
+from utils import utils,randomUtils,InputData, InputTypes
+#Internal Modules End--------------------------------------------------------------------------------
+
+class Importance(ForwardSampler):
+  """
+    MONTE CARLO Sampler
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:38]: Class docstrings need updating

File: framework/Samplers/Importance.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+"""
+  This module contains the Umbrella sampling strategy
+
+  Created on Feb 15, 2021
+  @author: Tanaya
+  supercedes Samplers.py from crisr
+"""
+#for future compatibility with Python 3--------------------------------------------------------------
+from __future__ import division, print_function, unicode_literals, absolute_import
+#End compatibility block for Python 3----------------------------------------------------------------
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:23]: This block is no longer needed since we dropped python 2 support.

File: framework/Samplers/Importance.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+  This module contains the Umbrella sampling strategy
+
+  Created on Feb 15, 2021
+  @author: Tanaya
+  supercedes Samplers.py from crisr
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:19]: Importance was not part of the original Samplers.py, so this line can be removed.

File: framework/Samplers/Importance.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        specifying input of cls.
+    """
+    inputSpecification = super(Importance, cls).getInputSpecification()
+    samplerInitInput = InputData.parameterInputFactory("samplerInit")
+    limit = InputData.parameterInputFactory("limit", contentType=InputTypes.IntegerType)
+    samplerInitInput.addSub(limit)
+    targetDistribution = InputData.parameterInputFactory("distribution")
+    samplerInitInput.addSub(targetDistribution)
+    samplingTypeInput = InputData.parameterInputFactory("samplingType", contentType=InputTypes.StringType)
+    samplerInitInput.addSub(samplingTypeInput)
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:57]: It's not a problem either way, but is there a reason you want to put the sampling type and distribution information inside the samplerInit rather than parallel with it?

Again, fine either way.

File: framework/Samplers/Importance.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    """
+      Provides the next sample to take.
+      After this method is called, the self.inputInfo should be ready to be sent
+      to the model
+      @ In, model, model instance, an instance of a model
+      @ In, myInput, list, a list of the original needed inputs for the model (e.g. list of files, etc.)
+      @ Out, None
+    """
+
+    uSample = self.distDict['uniform'].rvs()
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:98]: Is this a hard-coded key? I think usually the distDict is formed as `{variableName: distInstance}`; is `uniform` a variable name here?

File: framework/Samplers/Importance.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+    importanceWeight = self.distDict['target'].pdf(
+      importanceSample) / self.distDict['importance'].pdf(importanceSample)
+
+    self.inputInfo['SampledVars']['sample'] = importanceSample
+    self.inputInfo['ProbabilityWeight'] = 1.0
+    self.inputInfo['ProbabilityWeight-uniform'] = 1.0
+    self.inputInfo['ProbabilityWeight-target'] = self.distDict['target'].pdf(importanceSample)
+    self.inputInfo['ProbabilityWeight-importance'] = importanceWeight
+    self.inputInfo['PointProbability'] = importanceWeight
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:109]: Some of this meta information looks redundant to me. Is there a reason to keep the redundant entries?

File: tests/framework/ROM/TimeSeries/ARMA/basic.xml [hunk 19]
```diff
@@ -76,6 +76,9 @@
       <outTruncation domain="positive">Speed</outTruncation>
       <Fourier>604800, 302400, 86400, 43200</Fourier>
       <P>2</P>
+	  <!--<Fourier>604800, 302400, 86400, 43200, 720</Fourier>-->
```
Comment (PaulTalbot-INL) [inline tests/framework/ROM/TimeSeries/ARMA/basic.xml:79]: Did you mean to change this file?

File: tests/framework/ROM/TimeSeries/ARMA/Basic/dataSet_0.csv [hunk 20]
```diff
@@ -1,2018 +1,8929 @@
 Time,Speed
```
Comment (PaulTalbot-INL) [inline tests/framework/ROM/TimeSeries/ARMA/Basic/dataSet_0.csv:1]: Why is this file changed?

File: framework/Samplers/Importance.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+from operator import mul
+from functools import reduce
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from .ForwardSampler import ForwardSampler
+from utils import utils,randomUtils,InputData, InputTypes
+#Internal Modules End--------------------------------------------------------------------------------
+
+class Importance(ForwardSampler):
```
Comment (PaulTalbot-INL) [inline framework/Samplers/Importance.py:36]: Is there a regression test added for this Sampler?

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
A few more comments to consider; I did not look carefully at the umbrella sampler yet.

File: tests/framework/ROM/TimeSeries/ARMA/basic.xml (hunk 19 above)
Comment (nickblackmore) [reply tests/framework/ROM/TimeSeries/ARMA/basic.xml:79]: I think Tanaya changed that for experimentation and accidentally added it to tracked files. I'll change it back.

File: tests/framework/ROM/TimeSeries/ARMA/Basic/dataSet_0.csv (hunk 20 above)
Comment (nickblackmore) [reply tests/framework/ROM/TimeSeries/ARMA/Basic/dataSet_0.csv:1]: Same as the previous, I'll revert it to the previous state

File: framework/Samplers/Importance.py (hunk 18 above)
Comment (nickblackmore) [reply framework/Samplers/Importance.py:109]: I think we can remove 'PointProabaility'. Not sure if we actually need the ProbabilityWeight-target'. 'ProbabilityWeight-importance' is what's used to actually weight each sample.

File: framework/Samplers/Importance.py (hunk 21 above)
Comment (nickblackmore) [reply framework/Samplers/Importance.py:26]: Regression test has been added

File: tests/framework/test_Importance_Sampler.xml (hunk 11 above)
Comment (nickblackmore) [reply tests/framework/test_Importance_Sampler.xml:1]: Just wanted to double-check on this. I placed this test where the other sampler tests are located.

**Comment by aalfonsi** [comment]
@wangcj05 @PaulTalbot-INL do you want me to try to address the comments or you guys are taking care of this (since part of the IES collaboration with GWU)?

**Comment by PaulTalbot-INL** [comment]
Hi @aalfonsi,  @nickblackmore is the student at GWU working on this contract activity, so I think he'll be working on addressing comments. Certainly if there's something of particular note that you're interested in feel free to drop a note!

**Comment by aalfonsi** [comment]
> Hi @aalfonsi, @nickblackmore is the student at GWU working on this contract activity, so I think he'll be working on addressing comments. Certainly if there's something of particular note that you're interested in feel free to drop a note!
@PaulTalbot-INL great.... I am just trying to get rid of all my old branches and realized this PR is still pending... @nickblackmore let me know when and if I can delete the branch.

**Comment by nickblackmore** [comment]
@aalfonsi Hi, I think that I just need to resolve the conflicts-- I addressed the other requested changes. I've been working on another aspect of the project lately, but I can clean this up in the next couple of days.

**Comment by wangcj05** [comment]
Hi @PaulTalbot-INL @nickblackmore, do you have plan to finish this PR?

**Comment by aalfonsi** [comment]
@PaulTalbot-INL @nickblackmore @wangcj05 Is this going to be finalized? As mentioned in January 2022 I am trying to delete/clean up all my old branch. Thanks.

**Comment by aalfonsi** [comment]
@PaulTalbot-INL @nickblackmore @wangcj05 Closing this. It is clearly inactive since Jan 2021
